examples/vision/siamese.py

```diff
@@ -0,0 +1,232 @@
+
+"""
+# Siamese Network with Contrastive Loss
+
+Author: [Mehdi](https://github.com/s-mrb)<br>
+Date created: 2021/05/02<br>
+Last modified: 2020/05/03<br>
+Description: Similarity Learning Using Siamese Network with Contrastive Loss
+"""
+
+"""
+## Introduction
+
+The [Siamese Network](https://papers.nips.cc/paper/1993/file/288cc0ff022877bd3df94bc9360b9c5d-Paper.pdf)
+introduced by Bromley and LeCun to solve signature verification as an image
+matching problem is now widely used for finding the level of similarity between
+two images. This algorithm evolved to have many different forms, although
+the one we are going to use is Duplet Network. A duplet network involves
+two sister networks for finding embeddings/encodings and then a distance
+heuristic to find out how much these embeddings differ from each other.
+
+"""
+
+"""
+## Necessary imports
+"""
+
+import random
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras.models import Model
+from tensorflow.keras.layers.experimental.preprocessing import Rescaling
+from tensorflow.keras.layers import Flatten, Dense, Concatenate, Lambda, Input
+from tensorflow.keras.layers import Conv2D, Activation,AveragePooling2D
+from tensorflow.keras.datasets import mnist
+from tensorflow.keras.optimizers import RMSprop
+from tensorflow.keras import utils
+import matplotlib.pyplot as plt
+from keras import backend as K
+
+"""
+## Load data
+"""
+
+(x_train, y_train), (x_test, y_test) = mnist.load_data()
+
+# We change data type to float
+x_train = x_train.astype('float32')
+x_test = x_test.astype('float32')
+
+"""
+## Create pairs of images
+
+We want our model to be able to differentiate 0 from 1,2,3, .., 9 to do so we
+should teach it how each digit differs from each of the other digit, for this
+purpose we would select N random images from class A (lets say class of
+digit 0) and would pair it with N random images from class B (lets say class of
+digit 1), and would repeat this process untill class of digit 9. In this way we
+successfuly pair 0 with each other digit, now we would do the same for class of
+1,2,3 upto that of 9.
```

Comment:
Suggestion:

- Refactoring for improved syntax, grammar. Typo fix.
- 80 char per line
- Let me know if this reads OK @s-mrb 👍 

```suggestion
We will train the model to differentiate each digit from one another. For
example, digit `0` needs to be differentiated from the rest of the
digits (`1` through `9`), digit `1` - from `0` and `2` through `9`, and so on.
To carry this out, we will select N random images from class A (for example, for
digit `0`) and pair it with N random images from another class B (for example,
for digit `1`). Then, we can repeat this process for all classes of digits
(until digit `9`). Once we have paired digit `0` with other digits, we can
repeat this process for the remaining classes for the rest of the digits (from
`1` until `9`).
```

Reply:
This is great, Thanks.